### ISSUE TITLE
Replace Node interface with discriminated unions

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [microsoft/TypeScript#56275]
### 🔍 Search Terms

discrimnated type union node internal source api

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

Happy Halloween 🎃! In the spirit of the holiday, I'd like to resurrect #13634: using a discriminated unions for `Node` categories. Roughly paraphrasing #18285's very good description here:

As mentioned in https://github.com/Microsoft/TypeScript/pull/18214#issuecomment-326823448, TypeScript's source can remove a lot of casts by using a discriminated union for `Node`. 
While this should not affect TypeScript's runtime performance (as it just changes its types, mostly) it may cause TypeScript to take longer to compile itself.

#13634 was declined in 2018 for performance concerns (https://github.com/microsoft/TypeScript/issues/13634#issuecomment-413289616) but I hope that the years of performance improvements since then have since made this possible.

### 📃 Motivating Example

Switching TypeScript's `Node` from an interface from a discriminated type union would allow `.kind` checks to narrow types of values declared as `Node`.

```ts
import * as ts from "typescript";

declare const node: ts.Node;

if (node.kind === ts.SyntaxKind.FunctionDeclaration) {
  node.name?.text;
  //   ~~~~
  // Property 'name' does not exist on type 'Node'.
}
```

### 💻 Use Cases

In TypeScript itself, a match-case-and-full-word search for `/(?:name|node|parent) as \w{5,}/` in `src/**/*.ts` excluding `*test*` yields **1879** matches right now. After prototyping it locally and enabling [`@typescript-eslint/no-unnecessary-type-assertion`](https://typescript-eslint.io/rules/no-unnecessary-type-assertion), I mostly-automatically reduced it to **987** matches. More removals could likely be enacted with manual effort.
* Previous attempt: https://github.com/microsoft/TypeScript/pull/18285
* My attempt: https://github.com/Microsoft/TypeScript/pull/56274

In community projects that use the TypeScript AST, including [typescript-eslint](https://typescript-eslint.io), we'd similarly be able to have more precise type information on TS AST nodes.

#13634 mentions other consumers too, such as Babel's https://github.com/babel/babel/blob/c446ff85c28e117ebf3cd72cd34ef358f1077aa8/packages/babel-types/src/validators/is.ts.

**Comment by jakebailey** [comment]
@weswigham I know you were working on this somewhat recently too

**Comment by Andarist** [comment]
For reference: Wes' PR can be found [here](https://github.com/microsoft/TypeScript/pull/54148)

**Comment by weswigham** [comment]
Yeah, I have a recentish branch there, the 2.7x slowdown in compilation time and editor performance hurts though. Personally, I'm of the opinion that we should just eat the difference and use the worsened conditions as a forcing function for self-improvement, but that order of operations isn't a popular one :P

So... we *probably* can't take the change until we get the perf closer, which means looking into those traces @jakebailey took and trying to optimize those cases.
